Create sign-in/sign-up page and user page

weidak opened the pull request:
# Pull Request

## Description

Rectified issues

All components have no working backend. 
1. Created Sign-in/Sign-up components 
2. Created forgot password page
3. Created email verification page
4. Created profile update page

## Related Issue(s)

#11 

## Changes Made

Added login, verify and profile folders to /frontend/components/ to contain the components created.
Login:
- Login.tsx

Profle:
- ChangePassword.tsx
- Information.tsx
- Profile.tsx
- ProfileCard.tsx

Verify:
- Verify.tsx

Others:
- user.d.tsx (/types)
- loading.tsx (in /(pages)/ for verify, profile and login)
- error.tsx (in /(pages)/ for verify, profile and login)

## Screenshots (if applicable)

Sign-in:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/0e5b0ca1-8d5f-48d7-adca-a0ef5fd78188)

Forgot password: (After u key in email address and click submit)
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/7f31b79b-f710-493f-8842-89f82ebdf91d)

Loading: 
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/12d998aa-5710-473a-8103-3e1e23c93ac8)

Password visibility toggle:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/dff7fd0c-6c82-468c-8179-75fd3ad399b4)

Password not match:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/5e059cf2-31d4-4ad0-953e-438f95983d52)

Sign-Up:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/c1283a09-3c53-42cc-adb1-b045c394a531)

Verify:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/f3d985a5-2243-478d-b04c-3e5e03bfb9e8)

Profile:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/9a2d6889-92a9-4b7f-9da8-e89bbcc0109d)

Change Password:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/0daddc74-17fa-4166-951a-0c87ca08b7f7)

Clickable display pic:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/449e727d-69af-4749-8bfc-cbb67dd913cf)

Old and new passwords cannot be the same:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/d3587f23-a4d5-4be3-acd9-70f2c453b277)

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References
Please give any ideas below!

tryyang2001 commented on frontend/src/components/login/EyeHideLogo.tsx:
```diff
@@ -0,0 +1,39 @@
+const EyeHideLogo = ({ width = "100%", height = "100%" }: EyeHideLogoProps) => {
+  return (
+    <div style={{ width, height }}>
+      <svg
```
Can consider move this svg to `public` folder as that folder is meant to store all the images/assets.

tlyi replied:
@tryyang2001 I think in this case Weida has made the logo a component, so it cannot be placed inside `public` as the folder is meant to serve static assets only. 

@weidak I placed the PeerPrep logo in a component because I felt that it will be resized often throughout the pages so a component will be more convenient for reuse. However, I think we can consider treating svgs that are used only in limited pages/do not need to be resized often as just images instead and place them in the `public` folder to reduce the number of component files. What do you think?

tryyang2001 replied:
>  I think in this case Weida has made the logo a component, so it cannot be placed inside public as the folder is meant to serve static assets only.

Oh I thought the svg itself can be a static asset? We just import the image from the `public` folder to make our component. Does this make sense?

weidak replied:
> Oh I thought the svg itself can be a static asset? We just import the image from the `public` folder to make our component. Does this make sense?

Ok moved it to public, and replaced the logo components with <Image> that uses the svg in public folder.

tryyang2001 commented on frontend/src/components/login/Login.tsx:
```diff
@@ -0,0 +1,235 @@
+              setPassword(e.currentTarget.value);
+            }}
+            endContent={
+              isPasswordVisible ? (
```
Except for the different logos to render and a special horizontal padding "px-2" applied to the `EyeShow`, the two buttons seem to have the same properties. So one thing you can consider is to just put one button component here, and then move the ternary operator `isPasswordVisible` under the button component so that you only choose which logo to display and not which button component to render.

tryyang2001 commented on frontend/src/components/login/Login.tsx:
```diff
@@ -0,0 +1,235 @@
+                }}
+                placeholder="Name"
+              />
+              {password != checkPassword ? (
```
I will suggest we always use `!==` instead of `!=`. The difference is that operators like `!==` or `===` will not only check the value but also the type of the variables that you are comparing.

tryyang2001 commented on frontend/src/components/profile/ChangePassword.tsx:
```diff
@@ -0,0 +1,52 @@
+                        setConfirmNewPassword(e.currentTarget.value);
+                    }}
+                />
+                { (newPassword != confirmNewPassword) ? (<div className="text-center text-xs underline font-bold decoration-red-500">Passwords do not match</div>) : (<> </>)}
```
Same for this, can put `null` instead of the empty tag.

tryyang2001 commented on frontend/src/components/profile/ChangePassword.tsx:
```diff
@@ -0,0 +1,52 @@
+    const [confirmNewPassword, setConfirmNewPassword] = useState("");
+
+    return (
+        <>
```
I will suggest we put a `div` tag instead of empty tag for consistency.

tryyang2001 commented on frontend/src/components/profile/Information.tsx:
```diff
@@ -0,0 +1,46 @@
+
+export default function Information({setIsChangePassword}: InformationProps) {
+    return (
+    <>
```
Same for this, can consider changing to a `div` tag.

tryyang2001 reviewed (changes requested):
Since you mentioned that this will be a draft. I therefore converted this PR to draft first. When you think you are ready to merge to master, feel free to convert back.
Generally looks good to me, with some minor suggestions commented above. Feel free to take a look and see if would like to accept the suggestion. Thank you for making the changes, and also thanks for giving all the screenshots for the pages that you created!

tlyi commented:
I have some feedback/ considerations on the UI:
- White text on the `light-blue` background is a bit hard to read, you might want to consider using black text instead.
- I think in general, users would expect the buttons for the happy path to be on the right (e.g in Profile, the "Submit" button will be on the right instead of the "Change password" link), or at least that is my own opinion 😅 
- Can consider using a different style for error indication as the red underline feels like a typo... 
- Can consider using a different colour for hyperlinks for a more styled look. It would also be a good idea for us to standardise the colour of hyperlinks throughout the website, so maybe you can choose one and the rest of us will stick to it 😄 
- (Profile) How will the component react when the "Bio" input is too long? 

Other than these, most of it looks good to me! Thank you for making the pages! 😄

xingjie99 commented:
feedback from my side:

### UI: 
- Agree with what Le Yi said about the text color and background color and that the happy path should be on the right hand side

### Code:
- Instead of checking 
```
 {password != checkPassword ? (<div> Passwords do not match </div>) : (<></>)}
```
, 
would it be better to use set a flag and use `useEffect` hook to update the flag?

Eg

```
const [isEqual, setIsEqual] = useState(false);

useEffect(() => {
    if (password === checkPassword ) {
      setIsEqual(true);
    } else {
      setIsEqual(false);
    }
  }, [password, checkPassword ]);

...

{isEqual ? (<div> Passwords do not match </div>) : (<></>)}

```

I think doing this is trivial in this case but in the future when there is a more complicated condition to check for, handling the condition inside `useEffect` will make the code neater.

### Files:
- Just to clarify for the files `loading.tsx` and `error.tsx`, why are multiple of the same file created? Especially `loading.tsx`, shouldn't it be a loading component that is reusable by all pages (so should be extracted out to `/components`)

Anyway, everything else looks very good! Thanks for the hard work 👍👍

tryyang2001 commented:
If multiple `loading.tsx` file renders the same UI elements. We can put this `loading.tsx` in their parent directory. For example, in this case, you are displaying the spinner to display the loading state, you can actually put this file directly under the `/(pages)` directory.

> Just to clarify for the files loading.tsx and error.tsx, why are multiple of the same file created? Especially loading.tsx, shouldn't it be a loading component that is reusable by all pages (so should be extracted out to /components)

For @xingjie99 , I thought this `loading.tsx` and `error.tsx` are the file name required by NextJS as it automatically set up the routing for displaying all these elements, see this [docs](https://nextjs.org/docs/app/api-reference/file-conventions). So if I understand correctly, they can't be extracted out into components? Please correct me if I am wrong.

xingjie99 commented:
> If multiple `loading.tsx` file renders the same UI elements. We can put this `loading.tsx` in their parent directory. For example, in this case, you are displaying the spinner to display the loading state, you can actually put this file directly under the `/(pages)` directory.
> 
> > Just to clarify for the files loading.tsx and error.tsx, why are multiple of the same file created? Especially loading.tsx, shouldn't it be a loading component that is reusable by all pages (so should be extracted out to /components)
> 
> For @xingjie99 , I thought this `loading.tsx` and `error.tsx` are the file name required by NextJS as it automatically set up the routing for displaying all these elements, see this [docs](https://nextjs.org/docs/app/api-reference/file-conventions). So if I understand correctly, they can't be extracted out into components? Please correct me if I am wrong.

@tryyang2001 Oh.. I am not familiar with that perhaps if that is the case then can ignore my previous comment.

tryyang2001 commented:
> How will the component react when the "Bio" input is too long?

@tlyi Fyi I believe this should be handled by the backend. So as long as we only store the "bio" with appropriate length (I set the limit as 255 characters now), the frontend will never need to worry that they have a bio that is super long.

tlyi commented:
> > How will the component react when the "Bio" input is too long?
> 
> @tlyi Fyi I believe this should be handled by the backend. So as long as we only store the "bio" with appropriate length (I set the limit as 255 characters now), the frontend will never need to worry that they have a bio that is super long.

@tryyang2001 Oh, I meant to ask what would happen if it is longer than the length of the field in the component! Was wondering if the field would expand downwards or would it just scroll along the field (which might be quite inconvenient for the user).

weidak commented:
@tlyi 's comments:
> * White text on the `light-blue` background is a bit hard to read, you might want to consider using black text instead.
> * I think in general, users would expect the buttons for the happy path to be on the right (e.g in Profile, the "Submit" button will be on the right instead of the "Change password" link), or at least that is my own opinion 😅

Changed all button colors to blue to match Junsheng's implementation, moved buttons to the right (Example):
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/3ad00566-555c-4ae3-83f5-7796eaed1523)

> * Can consider using a different style for error indication as the red underline feels like a typo...

Ok changed to:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/02da4653-24de-47eb-b363-1a5aaa6eebfa)

> * Can consider using a different colour for hyperlinks for a more styled look. It would also be a good idea for us to standardise the colour of hyperlinks throughout the website, so maybe you can choose one and the rest of us will stick to it 😄

Im ok with this as its functional and looks like a link for people to press. "just move on" - cactus

> * (Profile) How will the component react when the "Bio" input is too long?

I just added in a limit for 50 chars, hopefully that works for u. I will sort more of the validation things (like email etc) once connected to the backend cos some coordination is required between db and here.
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/70256674/ac6ca245-83e8-4208-a6f5-218d531f53a4)

@xingjie99 :
> I think doing this is trivial in this case but in the future when there is a more complicated condition to check for, handling the condition inside useEffect will make the code neater.

Oh yeah I forgot about it.. thanks! Just move all the conditions already. 

thanks for the review everyone

weidak commented:
updated the PR description to match the changes made

tryyang2001 reviewed (approved):
LGTM, I have no more suggestion in my mind now. Thanks for updating!

c0j0s commented on frontend/public/submit_button.svg:
Would it be better for us to use an icon library instead? seems impractical to create an SVG file for every icon used.
